### PR TITLE
Intrepid2 develop

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
## Motivation

Fix typo of Devicetype to DeviceType.
Add HostBasis function interface. This is needed for orientation.

orientation tools are refactored. 

## Related Issues

#8310 

## Testing

tested on weaver

### CODE REVIEW

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp
Comment (mperego) [inline packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp:903]: This is something I wanted since some time, but it might be hard to implement  for hierarchical basis.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (mperego) [inline packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:452]: I remember Kara saying that `/* */` messes up with Doxygen. Have you checked whether this didn't disrupt Doxygen?

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (mperego) [inline packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:102]: @kyungjoo-kim we cannot do this. We are not in such a rush. Roger is working on including projections in Empire. There are tests using orientations and projections for hierarchical basis

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp
Comment (mperego) [inline packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp:206]: We should be able to make Orientation run on device without much effort. What issues did you encounter?

File: packages/intrepid2/unit-test/Orientation/Serial/CMakeLists.txt
Comment (mperego) [inline packages/intrepid2/unit-test/Orientation/Serial/CMakeLists.txt:74]: I'm against disabling tests

**Review by mperego — CHANGES_REQUESTED** [review]
I would expect that with a little more effort we should be able have orientations run without UVM. 
I don't think it is a good strategy to disable orientation tests for hierarchical basis. We are not in such a rush.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp:206]: This is initialization. There is no gain from exporting this to device.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:102]: I can include it but I would not test it.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp]: done.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp:903]: This is necessary in my opinion. We have frequent situation that a basis is templated. For computing reference (master) element operations, we probably want the host version.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp]: the headers are included but it does not effect anything. I don't understand why this is included in orientation tools if they are not used.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp
Comment (kyungjoo-kim) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp:903]: @CamelliaDPG @rppawlo I need this function in the hierarchical components as well to use it in the orientation tools.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
Comment (mperego) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp]: You are probably right. Maybe it was there before we started using `getReferenceSubCell` and it's no longer needed now.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp
Comment (mperego) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HCURL.hpp:206]: What problem do you encounter without UVM, if you try to run this on device?

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp:903]: @kyungjoo-kim @mperego Noted.  Yes, it will be slightly more complex for the tensor bases, but I think it won't be bad.  We will just need to get host versions of the component bases, and construct the host version of the composite basis using those.

**Comment by kyungjoo-kim** [comment]
@CamelliaDPG @mperego @rppawlo 

This PR is blocked by the issue of not having getHostBasis interface from the derived basis class. I tried to add the interface in the code. However, it is not that simple as some basis family functions use default exec space. The construction becomes tricky. I will leave this to @CamelliaDPG. This PR cannot be merged until then. Or I am going to submit another PR that excludes orientation tools.

**Comment by kyungjoo-kim** [comment]
close this for now and open #8774

**Pull request closed without merging**
